pkg/provider/pulumi/gcp/cloudrunner.go

```diff
@@ -0,0 +1,185 @@
+// Copyright Nitric Pty Ltd.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gcp
+
+import (
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi-gcp/sdk/v6/go/gcp/cloudrun"
+	"github.com/pulumi/pulumi-gcp/sdk/v6/go/gcp/projects"
+	"github.com/pulumi/pulumi-gcp/sdk/v6/go/gcp/pubsub"
+	"github.com/pulumi/pulumi-gcp/sdk/v6/go/gcp/serviceaccount"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+
+	"github.com/nitrictech/cli/pkg/provider/pulumi/common"
+	"github.com/nitrictech/cli/pkg/stack"
+	"github.com/nitrictech/cli/pkg/utils"
+)
+
+type CloudRunnerArgs struct {
+	Location  pulumi.StringInput
+	ProjectId string
+	Compute   stack.Compute
+	Image     *common.Image
+
+	Topics map[string]*pubsub.Topic
+}
+
+type CloudRunner struct {
+	pulumi.ResourceState
+
+	Name    string
+	Service *cloudrun.Service
+	Url     pulumi.StringInput
+}
+
+func (g *gcpProvider) newCloudRunner(ctx *pulumi.Context, name string, args *CloudRunnerArgs, opts ...pulumi.ResourceOption) (*CloudRunner, error) {
+	res := &CloudRunner{
+		Name: name,
+	}
+	err := ctx.RegisterComponentResource("nitric:func:GCPCloudRunner", name, res, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a service account for this cloud run instance
+	sa, err := serviceaccount.NewAccount(ctx, name+"-acct", &serviceaccount.AccountArgs{
+		AccountId: pulumi.String(utils.StringTrunc(name, 30-5) + "-acct"),
+	}, append(opts, pulumi.Parent(res))...)
+	if err != nil {
+		return nil, errors.WithMessage(err, "function serviceaccount "+name)
+	}
+
+	// Give project editor permissions
+	// FIXME: Trim this down
+	_, err = projects.NewIAMMember(ctx, name+"-editor", &projects.IAMMemberArgs{
```

Comment:
Should this be more granular with:
- roles/storage.objectAdmin
- roles/pubsub.editor
- roles/firestore.serviceAgent
- roles/secretmanager.secretAccessor

This will get replaced when the gcp least privilege implementation is done so it may not matter if it has the editor role for the time being.

Reply:
Otherwise, looks good :)

Reply:
Yeah, I agree it is not optimal. But good to do separately.